### PR TITLE
Sync & test levelbuilder toolbox options in applab/gamelab

### PR DESCRIPTION
**Pull request by islemaster**
For levelbuilder purposes, we keep a copy of our default toolbox JSON in the ruby model file which is used to populate the default toolbox JSON when creating a new level.  Unfortunately, we've been manually keeping these in sync (per a comment in applab/levels.js) and doing a poor job of it.

Here I resynchronize the toolbox JSON for applab and gamelab, and add dashboard tests that ensure they stay in sync.  Obviously covering this with tests instead of deduplicating it isn't ideal, but I'm not sure how to get our Rails app to efficiently read this value from a file, so this seems like a safe stopgap measure.

When the test fails it will report which palette keys are missing. Example:

```
Minitest::Assertion: Ruby-JavaScript palette mismatch
'getAlpha' missing from applab/levels.js
'getAlpa' missing from applab.rb
```

### CODE REVIEW

**Comment by islemaster** [comment]
CC @laurelfan @Bjvanminnen

File: apps/src/gamelab/levels.js [hunk 1]
```diff
@@ -215,7 +215,7 @@ levels.custom = utils.extend(levels.sandbox, {
     "functionParams_n": null,
     "callMyFunction": null,
     "callMyFunction_n": null,
-    "return": null,
+    "return": null
```
Comment (islemaster) [inline apps/src/gamelab/levels.js:218]: This object must be valid JSON (not just JavaScript) for the test to work.  Omitting trailing commas seems like a small price to pay to cover this.

File: dashboard/test/models/gamelab_test.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    from_levels_js.keys.each do |key|
+      errors.push("Palette entry '#{key}' missing from gamelab.rb") unless from_gamelab_rb.key? key
+    end
+    from_gamelab_rb.keys.each do |key|
+      errors.push("codeFunctions entry '#{key}' missing from gamelab/levels.js") unless from_levels_js.key? key
+    end
+    assert errors.empty?, "Ruby-JavaScript palette mismatch:\n#{errors.join("\n")}"
+  end
+
+end
```
Comment (Bjvanminnen) [inline dashboard/test/models/gamelab_test.rb:24]: This is cool! Would it make more sense to share more code for these two, considering they're doing basically the same thing, just with different paths?

Comment (islemaster) [inline dashboard/test/models/gamelab_test.rb:24]: Probably.  Maybe the following helpers?
- Something to help with retrieving the block from the JavaScript file
- An `assert_same_keys` helper of some kind

Comment (islemaster) [inline dashboard/test/models/gamelab_test.rb:24]: I have to admit, my Ruby is so rusty that I'm not sure where such a helper would live.

**Comment by islemaster** [comment]
Extracted `assert_same_keys`.  The regex to grab the right stuff from the JS file is so specific (and so far from working for any general case) that I'm tempted to just leave it duplicated - I don't think we're likely to reuse it anywhere else soon.

**Comment by aoby** [comment]
lgtm

If we wanted to actually dedupe instated of test, we could put the json in its own file in a common location and load from both .js and .rb.

File: dashboard/test/test_helper.rb [hunk 3]
```diff
@@ -161,6 +161,20 @@ def assert_no_change(expressions, message = nil, &block)
       assert_equal(before[i], e.call, error)
     end
   end
+
+  # Given two hashes, ensure that they have the same set of keys in both
+  # directions, collecting up errors so we can pretty-print them all in one go
+  # if any errors are found.
+  def assert_same_keys(a, b, a_name = 'A', b_name = 'B', message = nil)
```
Comment (laurelfan) [inline dashboard/test/test_helper.rb:168]: `-` for arrays does what you would want it to, so you could just do something like::

```
assert_equal a.keys, b.keys,  
"#{message}\nkeys in A but not B: #{a.keys - b.keys} \n keys in B but not A: #{b.keys - a.keys}"
```

or if you don't also want the "Expected:" and "Actual:" lines from assert_equal you could 

```
if a.keys != b.keys
  assert false, "#{message...
end
```

Comment (islemaster) [inline dashboard/test/test_helper.rb:168]: o_O what is this dark magic?

Updating implementation in a new PR.

**Pull request merged**